api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -130,13 +130,13 @@ public static Schema select(Schema schema, Set<Integer> fieldIds) {
   public static Types.StructType selectNot(Types.StructType struct, Set<Integer> fieldIds) {
     Set<Integer> projectedIds = getIdsInternal(struct);
     projectedIds.removeAll(fieldIds);
-    return select(struct, projectedIds);
+    return project(struct, projectedIds);
```

Comment:
One issue here is selectNot doest not actually deselect children when a parent ID is not selected. Previously this is because getProjectedIDs (behind getIdsInternal) would not return parent struct ids, so removing it from the set of projectedIds would not do anything.

Now It will not work because removing a parentID still leaves all child IDs. We could fix this but it would be a change in behavior from the previous code.

Reply:
I think I agree with the decision to not change the behavior of this method, even though the opposite of "select" behavior would be to fully remove a struct when its ID is passed in `fieldIds`.

But I don't think that `project` is quite correct either. Consider the example schema `1: id bigint, 2: location struct<3: lat double, 4: long double>`. Previously, `selectNot(t, set(3, 4))` would produce `1: id bigint` and omit the location entirely. Using project with the updated `GetProjectedIds`, the projected ID set will be {1, 2, 3, 4} and not `{1, 3, 4}`. That would result in the same call producing `1: id bigint, 2: location struct<>`, which introduces a new bug because now there is an unexpected extra field.

To clean this up, I think we need a version of `GetProjectedIds` that doesn't select structs and uses the old behavior.

Reply:
That seems like the right behavior to me? Shouldn't you be required to explicitly omit the parent if you don't want the that element? Otherwise there would be no way to "selectNot"  and only get back the empty struct.

Reply:
Wrote up these test cases, i'll run the full test suite to make sure this works with our other usages

```
    Schema schema = new Schema(
        Lists.newArrayList(
            required(1, "id", Types.LongType.get()),
            required(2, "location", Types.StructType.of(
                required(3, "lat", Types.DoubleType.get()),
                required(4, "long", Types.DoubleType.get())
            ))));

    Schema expectedNoPrimitive = new Schema(
        Lists.newArrayList(
            required(2, "location", Types.StructType.of(
                required(3, "lat", Types.DoubleType.get()),
                required(4, "long", Types.DoubleType.get())
            ))));

    Schema actualNoPrimitve = TypeUtil.selectNot(schema, Sets.newHashSet(1));
    Assert.assertEquals(expectedNoPrimitive.asStruct(), actualNoPrimitve.asStruct());

    // Expected legacy behavior is to completely remove structs if their elements are removed
    Schema expectedNoStructElements = new Schema(required(1, "id", Types.LongType.get()));
    Schema actualNoStructElements = TypeUtil.selectNot(schema, Sets.newHashSet(3, 4));
    Assert.assertEquals(expectedNoStructElements.asStruct(), actualNoStructElements.asStruct());

    // Expected legacy behavior is to ignore selectNot on struct elements.
    Schema actualNoStruct = TypeUtil.selectNot(schema, Sets.newHashSet(2));
    Assert.assertEquals(schema.asStruct(), actualNoStruct.asStruct());
    ```